tests/test_slots.py

```diff
@@ -325,76 +325,86 @@ class C2(C1Bare):
-@pytest.mark.parametrize("slots", [True, False])
-def test_closure_cell_rewriting_cls_static(slots):
-    """
-    Slot classes support proper closure cell rewriting for class- and static
-    methods.
-    """
-    # Python can reuse closure cells, so we create new classes just for
-    # this test.
-
-    @attr.s(slots=slots)
-    class C:
-        @classmethod
-        def clsmethod(cls):
-            return __class__  # noqa: F821
-
-    assert C.clsmethod() is C
-
-    @attr.s(slots=slots)
-    class D:
-        @staticmethod
-        def statmethod():
-            return __class__  # noqa: F821
-
-    assert D.statmethod() is D
+class TestClosureCellRewriting(object):
+    def test_closure_cell_rewriting(self):
+        """
+        Slot classes support proper closure cell rewriting.
+
+        This affects features like `__class__` and the no-arg super().
+        """
+        non_slot_instance = C1(x=1, y="test")
+        slot_instance = C1Slots(x=1, y="test")
+
+        assert non_slot_instance.my_class() is C1
+        assert slot_instance.my_class() is C1Slots
+
+        # Just assert they return something, and not an exception.
+        assert non_slot_instance.my_super()
+        assert slot_instance.my_super()
+
+    def test_inheritance(self):
+        """
+        Slot classes support proper closure cell rewriting when inheriting.
+
+        This affects features like `__class__` and the no-arg super().
+        """
+        @attr.s
+        class C2(C1):
+            def my_subclass(self):
+                return __class__  # NOQA: F821
+
+        @attr.s
+        class C2Slots(C1Slots):
+            def my_subclass(self):
+                return __class__  # NOQA: F821
+
+        non_slot_instance = C2(x=1, y="test")
+        slot_instance = C2Slots(x=1, y="test")
+
+        assert non_slot_instance.my_class() is C1
+        assert slot_instance.my_class() is C1Slots
+
+        # Just assert they return something, and not an exception.
+        assert non_slot_instance.my_super()
+        assert slot_instance.my_super()
+
+        assert non_slot_instance.my_subclass() is C2
+        assert slot_instance.my_subclass() is C2Slots
+
+    @pytest.mark.parametrize("slots", [True, False])
+    def test_cls_static(self, slots):
+        """
+        Slot classes support proper closure cell rewriting for class- and
+        static methods.
+        """
+        # Python can reuse closure cells, so we create new classes just for
+        # this test.
+
+        @attr.s(slots=slots)
+        class C:
+            @classmethod
+            def clsmethod(cls):
+                return __class__  # noqa: F821
+
+        assert C.clsmethod() is C
+
+        @attr.s(slots=slots)
+        class D:
+            @staticmethod
+            def statmethod():
+                return __class__  # noqa: F821
+
+        assert D.statmethod() is D
+
+    @pytest.mark.skipif(
+        PYPY,
+        reason="ctypes are used only on CPython 3"
```

Comment:
CPython 2, right?

Reply:
cpython2 does not support argument-less super, so its not necessary there

Reply:
Well, on py2 the whole code path is completely unused.  We just didn’t anticipate ctypes-free systems so we didn’t put any effort into masking it out.